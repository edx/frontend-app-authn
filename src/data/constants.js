// URL Paths
export const LOGIN_PAGE = '/login';
export const REGISTER_PAGE = '/register';
export const RESET_PAGE = '/reset';
export const DEFAULT_REDIRECT_URL = '/dashboard';
export const PASSWORD_RESET_CONFIRM = '/password_reset_confirm/:token/';
export const PAGE_NOT_FOUND = '/notfound';

// Stateful Submit Button States
export const DEFAULT_STATE = 'default';
export const PENDING_STATE = 'pending';

// Constants
export const SUPPORTED_ICON_CLASSES = ['apple', 'facebook', 'google', 'microsoft'];
export const REGISTRATION_VALIDITY_MAP = {};
export const REGISTRATION_OPTIONAL_MAP = {};
export const REGISTRATION_EXTRA_FIELDS = [
  'confirm_email',
  'level_of_education',
  'gender',
  'year_of_birth',
  'mailing_address',
  'goals',
  'honor_code',
  'terms_of_service',
  'city',
  'country',
];

export const VALID_EMAIL_REGEX = '(^[-!#$%&\'*+/=?^_`{}|~0-9A-Z]+(\\.[-!#$%&\'*+/=?^_`{}|~0-9A-Z]+)*'
                                 + '|^"([\\001-\\010\\013\\014\\016-\\037!#-\\[\\]-\\177]|\\\\[\\001-\\011\\013\\014\\016-\\177])*"'
                                 + ')@((?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\\.)+)(?:[A-Z0-9-]{2,63})'
                                 + '|\\[(25[0-5]|2[0-4]\\d|[0-1]?\\d?\\d)(\\.(25[0-5]|2[0-4]\\d|[0-1]?\\d?\\d)){3}\\]$';
