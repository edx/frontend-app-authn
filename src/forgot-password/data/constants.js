export default 'internal-server-error';
