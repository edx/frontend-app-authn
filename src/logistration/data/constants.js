export const INACTIVE_USER = 'inactive-user';
export const INTERNAL_SERVER_ERROR = 'internal-server-error';
export const NON_COMPLIANT_PASSWORD_EXCEPTION = 'NonCompliantPasswordException';
