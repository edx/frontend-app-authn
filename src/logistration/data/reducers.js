import {
  REGISTER_NEW_USER,
  LOGIN_REQUEST,
  THIRD_PARTY_AUTH_CONTEXT,
} from './actions';

export const defaultState = {
  registrationResult: {},
  loginResult: {},
};

const reducer = (state = defaultState, action) => {
  switch (action.type) {
    case REGISTER_NEW_USER.BEGIN:
      return {
        ...state,
      };
    case REGISTER_NEW_USER.SUCCESS:
      return {
        ...state,
        registrationResult: action.payload,
      };
    case REGISTER_NEW_USER.FAILURE:
      return {
        ...state,
      };
    case LOGIN_REQUEST.BEGIN:
      return {
        ...state,
      };
    case LOGIN_REQUEST.SUCCESS:
      return {
        ...state,
        loginResult: action.payload,
      };
    case LOGIN_REQUEST.FAILURE:
      return {
        ...state,
      };
    case THIRD_PARTY_AUTH_CONTEXT.BEGIN:
      return {
        ...state,
      };
    case THIRD_PARTY_AUTH_CONTEXT.SUCCESS:
      return {
        ...state,
        thirdPartyAuthContext: action.payload.thirdPartyAuthContext,
      };
    case THIRD_PARTY_AUTH_CONTEXT.FAILURE:
      return {
        ...state,
      };
    default:
      return state;
  }
};

export default reducer;
