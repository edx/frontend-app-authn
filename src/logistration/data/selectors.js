import { createSelector } from 'reselect';

export const storeName = 'logistration';

export const logistrationSelector = state => ({ ...state[storeName] });

export const loginRequestSelector = createSelector(
  logistrationSelector,
  logistration => logistration.loginResult,
);

export const loginErrorSelector = createSelector(
  logistrationSelector,
  logistration => logistration.loginError,
);

export const registrationRequestSelector = createSelector(
  logistrationSelector,
  logistration => logistration.registrationResult,
);

export const thirdPartyAuthContextSelector = createSelector(
  logistrationSelector,
  logistration => logistration.thirdPartyAuthContext,
);
