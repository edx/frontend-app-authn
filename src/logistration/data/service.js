import { getConfig } from '@edx/frontend-platform';
import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
import querystring from 'querystring';

export async function postNewUser(registrationInformation) {
  const requestConfig = {
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    isPublic: true,
  };

  const { data } = await getAuthenticatedHttpClient()
    .post(
      `${getConfig().LMS_BASE_URL}/user_api/v1/account/registration/`,
      querystring.stringify(registrationInformation),
      requestConfig,
    )
    .catch((e) => {
      throw (e);
    });

  return data;
}

export async function login(creds) {
  const requestConfig = {
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    isPublic: true,
  };

  const { data } = await getAuthenticatedHttpClient()
    .post(
      `${getConfig().LMS_BASE_URL}/login_ajax`,
      querystring.stringify(creds),
      requestConfig,
    )
    .catch((e) => {
      throw (e);
    });

  return {
    redirectUrl: data.redirect_url || '/dashboard',
    success: data.success || false,
  };
}
