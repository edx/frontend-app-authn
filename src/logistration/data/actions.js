import { AsyncActionType } from '../../data/utils';

export const REGISTER_NEW_USER = new AsyncActionType('REGISTRATION', 'REGISTER_NEW_USER');
export const LOGIN_REQUEST = new AsyncActionType('LOGIN', 'REQUEST');
export const THIRD_PARTY_AUTH_CONTEXT = new AsyncActionType('THIRD_PARTY_AUTH', 'GET_THIRD_PARTY_AUTH_CONTEXT');

// Register

export const registerNewUser = registrationInfo => ({
  type: REGISTER_NEW_USER.BASE,
  payload: { registrationInfo },
});

export const registerNewUserBegin = () => ({
  type: REGISTER_NEW_USER.BEGIN,
});

export const registerNewUserSuccess = (redirectUrl, success) => ({
  type: REGISTER_NEW_USER.SUCCESS,
  payload: { redirectUrl, success },
});

export const registerNewUserFailure = () => ({
  type: REGISTER_NEW_USER.FAILURE,
});

// Login
export const loginRequest = creds => ({
  type: LOGIN_REQUEST.BASE,
  payload: { creds },
});

export const loginRequestBegin = () => ({
  type: LOGIN_REQUEST.BEGIN,
});

export const loginRequestSuccess = (redirectUrl, success) => ({
  type: LOGIN_REQUEST.SUCCESS,
  payload: { redirectUrl, success },
});

export const loginRequestFailure = () => ({
  type: LOGIN_REQUEST.FAILURE,
});

// Third party auth context
export const getThirdPartyAuthContext = (urlParams) => ({
  type: THIRD_PARTY_AUTH_CONTEXT.BASE,
  payload: { urlParams },
});

export const getThirdPartyAuthContextBegin = () => ({
  type: THIRD_PARTY_AUTH_CONTEXT.BEGIN,
});

export const getThirdPartyAuthContextSuccess = (thirdPartyAuthContext) => ({
  type: THIRD_PARTY_AUTH_CONTEXT.SUCCESS,
  payload: { thirdPartyAuthContext },
});

export const getThirdPartyAuthContextFailure = () => ({
  type: THIRD_PARTY_AUTH_CONTEXT.FAILURE,
});
