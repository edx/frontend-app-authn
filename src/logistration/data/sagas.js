import { call, put, takeEvery } from 'redux-saga/effects';

import { camelCaseObject } from '@edx/frontend-platform';
import { logError } from '@edx/frontend-platform/logging';
import { INTERNAL_SERVER_ERROR } from './constants';

// Actions
import {
  REGISTER_NEW_USER,
  registerNewUserBegin,
  registerNewUserFailure,
  registerNewUserSuccess,
  LOGIN_REQUEST,
  loginRequestBegin,
  loginRequestFailure,
  loginRequestSuccess,
  REGISTER_FORM_VALIDATIONS,
  fetchRealtimeValidationsBegin,
  fetchRealtimeValidationsSuccess,
  fetchRealtimeValidationsFailure,
  THIRD_PARTY_AUTH_CONTEXT,
  getThirdPartyAuthContextBegin,
  getThirdPartyAuthContextSuccess,
  getThirdPartyAuthContextFailure,
  REGISTER_FORM,
  fetchRegistrationFormBegin,
  fetchRegistrationFormSuccess,
  fetchRegistrationFormFailure,
} from './actions';

// Services
import {
  getFieldsValidations,
  getRegistrationForm,
  getThirdPartyAuthContext,
  registerRequest,
  loginRequest,
} from './service';

export function* handleNewUserRegistration(action) {
  try {
    yield put(registerNewUserBegin());

    const { redirectUrl, success } = yield call(registerRequest, action.payload.registrationInfo);

    yield put(registerNewUserSuccess(
      redirectUrl,
      success,
    ));
  } catch (e) {
    const statusCodes = [400, 409, 403];
    if (e.response && statusCodes.includes(e.response.status)) {
      yield put(registerNewUserFailure(e.response.data));
    }
    logError(e);
  }
}

export function* handleLoginRequest(action) {
  try {
    yield put(loginRequestBegin());

    const { redirectUrl, success } = yield call(loginRequest, action.payload.creds);

    yield put(loginRequestSuccess(
      redirectUrl,
      success,
    ));
  } catch (e) {
    const statusCodes = [400];
    if (e.response) {
      if (statusCodes.includes(e.response.status)) {
        yield put(loginRequestFailure(camelCaseObject(e.response.data)));
      } else {
        yield put(loginRequestFailure(camelCaseObject({ errorCode: INTERNAL_SERVER_ERROR })));
      }
    }
    logError(e);
  }
}

export function* fetchThirdPartyAuthContext(action) {
  try {
    yield put(getThirdPartyAuthContextBegin());
    const { thirdPartyAuthContext } = yield call(getThirdPartyAuthContext, action.payload.urlParams);

    yield put(getThirdPartyAuthContextSuccess(
      thirdPartyAuthContext,
    ));
  } catch (e) {
    yield put(getThirdPartyAuthContextFailure());
    logError(e);
  }
}

export function* fetchRegistrationForm() {
  try {
    yield put(fetchRegistrationFormBegin());
    const { registrationForm } = yield call(getRegistrationForm);

    yield put(fetchRegistrationFormSuccess(
      registrationForm,
    ));
  } catch (e) {
    yield put(fetchRegistrationFormFailure());
    logError(e);
  }
}

export function* fetchRealtimeValidations(action) {
  try {
    yield put(fetchRealtimeValidationsBegin());
    const { fieldValidations } = yield call(getFieldsValidations, action.payload.formPayload);

    yield put(fetchRealtimeValidationsSuccess(
      fieldValidations,
    ));
  } catch (e) {
    yield put(fetchRealtimeValidationsFailure());
    logError(e);
  }
}

export default function* saga() {
  yield takeEvery(REGISTER_NEW_USER.BASE, handleNewUserRegistration);
  yield takeEvery(LOGIN_REQUEST.BASE, handleLoginRequest);
  yield takeEvery(THIRD_PARTY_AUTH_CONTEXT.BASE, fetchThirdPartyAuthContext);
  yield takeEvery(REGISTER_FORM.BASE, fetchRegistrationForm);
  yield takeEvery(REGISTER_FORM_VALIDATIONS.BASE, fetchRealtimeValidations);
}
