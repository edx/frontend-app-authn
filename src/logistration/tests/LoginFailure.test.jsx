import React from 'react';
import renderer from 'react-test-renderer';
import { IntlProvider } from '@edx/frontend-platform/i18n';

import LoginFailureMessage from '../LoginFailure';
import { INACTIVE_USER, INTERNAL_SERVER_ERROR, NON_COMPLIANT_PASSWORD_EXCEPTION } from '../data/constants';

describe('LoginFailureMessage', () => {
  let props = {};

  it('should match non compliant password error message snapshot', () => {
    props = {
      loginError: {
        errorCode: NON_COMPLIANT_PASSWORD_EXCEPTION,
      },
    };

    const tree = renderer.create(
      <IntlProvider locale="en">
        <LoginFailureMessage {...props} />
      </IntlProvider>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should match inactive user error message snapshot', () => {
    props = {
      loginError: {
        email: 'text@example.com',
        errorCode: INACTIVE_USER,
        context: {
          platformName: 'openedX',
          supportLink: 'https://support.edx.org/',
        },
      },
    };

    const tree = renderer.create(
      <IntlProvider locale="en">
        <LoginFailureMessage {...props} />
      </IntlProvider>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should match internal server error message snapshot', () => {
    props = {
      loginError: {
        errorCode: INTERNAL_SERVER_ERROR,
      },
    };

    const tree = renderer.create(
      <IntlProvider locale="en">
        <LoginFailureMessage {...props} />
      </IntlProvider>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should match direct render of error message snapshot', () => {
    props = {
      loginError: {
        value: 'Email or password is incorrect.',
      },
    };

    const tree = renderer.create(
      <IntlProvider locale="en">
        <LoginFailureMessage {...props} />
      </IntlProvider>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should match error message containing link snapshot', () => {
    props = {
      loginError: {
        value: 'To be on the safe side, you can reset your password <a href="/reset">here</a> before you try again.\n',
      },
    };

    const tree = renderer.create(
      <IntlProvider locale="en">
        <LoginFailureMessage {...props} />
      </IntlProvider>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
