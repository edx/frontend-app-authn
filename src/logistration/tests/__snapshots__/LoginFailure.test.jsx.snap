// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginFailureMessage should match direct render of error message snapshot 1`] = `
<div
  className="fade alert alert-danger show"
  role="alert"
>
  <div
    className="alert-heading h4"
  >
    <span>
      We couldn't sign you in.
    </span>
  </div>
  <ul>
    <li>
      Email or password is incorrect.
    </li>
  </ul>
</div>
`;

exports[`LoginFailureMessage should match error message containing link snapshot 1`] = `
<div
  className="fade alert alert-danger show"
  role="alert"
>
  <div
    className="alert-heading h4"
  >
    <span>
      We couldn't sign you in.
    </span>
  </div>
  <ul>
    <li>
      To be on the safe side, you can reset your password 
      <a
        href="/reset"
        onClick={[Function]}
        target="_self"
      >
        here
      </a>
       before you try again.
    </li>
  </ul>
</div>
`;

exports[`LoginFailureMessage should match inactive user error message snapshot 1`] = `
<div
  className="fade alert alert-danger show"
  role="alert"
>
  <div
    className="alert-heading h4"
  >
    <span>
      We couldn't sign you in.
    </span>
  </div>
  <ul>
    <li>
      <span>
        In order to sign in, you need to activate your account.
        <br />
         
        <br />
        We just sent an activation link to 
        <strong>
          text@example.com
        </strong>
        . If you do not receive an email, check your spam folders or 
        <a
          className="alert-link"
          href="https://support.edx.org/"
          onClick={[Function]}
          onKeyDown={[Function]}
        >
          <span>
            contact openedX Support
          </span>
        </a>
        .
      </span>
    </li>
  </ul>
</div>
`;

exports[`LoginFailureMessage should match internal server error message snapshot 1`] = `
<div
  className="fade alert alert-danger show"
  role="alert"
>
  <div
    className="alert-heading h4"
  >
    <span>
      We couldn't sign you in.
    </span>
  </div>
  <ul>
    <li>
      <span>
        An error has occurred. Try refreshing the page, or check your Internet connection.
      </span>
    </li>
  </ul>
</div>
`;

exports[`LoginFailureMessage should match non compliant password error message snapshot 1`] = `
<div
  className="fade alert alert-danger show"
  role="alert"
>
  <div
    className="alert-heading h4"
  >
    <span>
      We couldn't sign you in.
    </span>
  </div>
  <ul>
    <li>
      <span>
        <strong>
          We recently changed our password requirements
        </strong>
         
        <br />
        Your current password does not meet the new security requirements. We just sent a password-reset message to the email address associated with this account. Thank you for helping us keep your data safe.
      </span>
    </li>
  </ul>
</div>
`;
